clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -0,0 +1,289 @@
+  //          + 1, if all B' branch to B on a condition of the form "if *p".
+  //            0, otherwise.
+
+  // DeclMap stores the number of preds of EB which branch to EB on a condition
+  // like "if *p". If all the preds of EB do this then we can increment
+  // EB->Gen[D] by 1.
+  using DeclMapTy = llvm::DenseMap<const VarDecl *, unsigned>;
+  DeclMapTy DeclMap;
+
+  BoundsMapTy Intersections;
+  bool ItersectionEmpty = true;
+
+  for (const CFGBlock *pred : EB->Block->preds()) {
+    if (!BlockMap.count(pred))
+      continue;
+    auto PredEB = BlockMap[pred];
+
+    if (ItersectionEmpty) {
+      Intersections = PredEB->Gen;
+      ItersectionEmpty = false;
+    } else
+      Intersections = Intersect(Intersections, PredEB->Gen);
+
+    // Fill DeclMap.
+    if (const Expr *E = GetTerminatorCondition(pred)) {
+      if (!ContainsPointerDeref(E))
+        continue;
+
+      // We can update EB->Gen only if EB is the then block of the condition
+      // "if *p". For example:
+
+      // B1: if (*p)
+      // B2:   foo();
+      // B3: else bar();
+
+      //      B1
+      //     /  \
+      //    B2   B3
+
+      // B1: preds[], succs[B2, B3]
+      // B2: preds[B1], succs[]
+      // B3: preds[B1], succs[]
+
+      // This means for an if condition, the then block is always the first
+      // block in the list of succs. We use this fact to check if EB is the
+      // then block of the if condition.
+      if (const auto *I = pred->succs().begin())
+        if (*I != EB->Block)
+          continue;
+
+      const VarDecl *D = GetVarDecl(E);
+      if (D && D->getType()->isCheckedPointerNtArrayType()) {
+        if (!DeclMap.count(D))
+          DeclMap[D] = 1;
+        else
+          DeclMap[D]++;
+      }
+    }
+  }
+  EB->Gen = Intersections;
+
+  // Now update EB->Gen.
+  for (const auto item : DeclMap) {
+    unsigned count = item.second;
+    if (count == EB->Block->pred_size()) {
+      const auto *D = item.first;
+      if (!EB->Gen.count(D))
+        EB->Gen[D] = 1;
+      else
+        EB->Gen[D]++;
+    }
+  }
+}
+
+void BoundsAnalysis::UpdateInMap(ElevatedCFGBlock *EB, BlockMapTy BlockMap) {
+  // In(B) = min(n Out(B')), where B E preds(B).
+
+  // In(B) is the intersection of the Out's of all preds of B. For the
+  // intersection, we pick the Decl with the smaller upper bound. For example,
+  // if preds(B) = {X, Y} and Out(X)= {p:1, q:2, r:0} and Out(Y) = {p:0, q:3,
+  // s:2} then In(B) = {p:0, q:2}.
+  BoundsMapTy Intersections;
+  bool ItersectionEmpty = true;
+
+  for (const CFGBlock *pred : EB->Block->preds()) {
+    if (!BlockMap.count(pred))
+      continue;
+    auto PredEB = BlockMap[pred];
+
+    if (ItersectionEmpty) {
+      Intersections = PredEB->Out;
+      ItersectionEmpty = false;
+    } else
+      Intersections = Intersect(Intersections, PredEB->Out);
+  }
+  EB->In = Intersections;
+}
+
+BoundsMapTy BoundsAnalysis::UpdateOutMap(ElevatedCFGBlock *EB) {
+  // Out(B) = max(Gen(B) u In(B)).
```

Comment:
If you reformulate Gen to be "an expression of the form *e" known to be non-zero", then you could walk over all the bounds in In and widen them if possible.  Specifically, for any variable v in In, if the bounds of v have the form (low, high + i) and e has the form (high + i), then the bounds can widened by 1.